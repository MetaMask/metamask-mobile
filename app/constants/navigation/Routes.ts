const Routes = {
  WALLET_VIEW: 'WalletView',
  BROWSER_TAB_HOME: 'BrowserTabHome',
  BROWSER_URL_MODAL: 'BrowserUrlModal',
  BROWSER_VIEW: 'BrowserView',
  FIAT_ON_RAMP_AGGREGATOR: {
    ID: 'FiatOnRampAggregator',
    GET_STARTED: 'GetStarted',
    PAYMENT_METHOD: 'PaymentMethod',
    PAYMENT_METHOD_HAS_STARTED: 'PaymentMethodHasStarted',
    AMOUNT_TO_BUY: 'AmountToBuy',
    GET_QUOTES: 'GetQuotes',
    CHECKOUT: 'Checkout',
    REGION: 'Region',
    REGION_HAS_STARTED: 'RegionHasStarted',
    ORDER_DETAILS: 'OrderDetails',
    SETTINGS: 'OnRampSettings',
    ADD_ACTIVATION_KEY: 'OnrampAddActivationKey',
  },
  QR_SCANNER: 'QRScanner',
  TRANSACTIONS_VIEW: 'TransactionsView',
  MODAL: {
    DELETE_WALLET: 'DeleteWalletModal',
    ROOT_MODAL_FLOW: 'RootModalFlow',
    MODAL_CONFIRMATION: 'ModalConfirmation',
    WHATS_NEW: 'WhatsNewModal',
    TURN_OFF_REMEMBER_ME: 'TurnOffRememberMeModal',
    UPDATE_NEEDED: 'UpdateNeededModal',
    ENABLE_AUTOMATIC_SECURITY_CHECKS: 'EnableAutomaticSecurityChecksModal',
    DETECTED_TOKENS: 'DetectedTokens',
    SRP_REVEAL_QUIZ: 'SRPRevealQuiz',
  },
  ONBOARDING: {
    ROOT_NAV: 'OnboardingRootNav',
    HOME_NAV: 'HomeNav',
    ONBOARDING: 'Onboarding',
    LOGIN: 'Login',
    NAV: 'OnboardingNav',
    MANUAL_BACKUP: {
      STEP_3: 'ManualBackupStep3',
    },
  },
  SEND_FLOW: {
    SEND_TO: 'SendTo',
  },
  ACCOUNT_BACKUP: {
    STEP_1_B: 'AccountBackupStep1B',
  },
  SETTINGS: {
    CONTACT_FORM: 'ContactForm',
    REVEAL_PRIVATE_CREDENTIAL: 'RevealPrivateCredentialView',
  },
  SHEET: {
    ACCOUNT_SELECTOR: 'AccountSelector',
    ACCOUNT_CONNECT: 'AccountConnect',
    ACCOUNT_PERMISSIONS: 'AccountPermissions',
  },
  BROWSER: {
    HOME: 'BrowserTabHome',
    URL_MODAL: 'BrowserUrlModal',
    VIEW: 'BrowserView',
  },
  WEBVIEW: {
    MAIN: 'Webview',
    SIMPLE: 'SimpleWebview',
  },
  WALLET: {
    HOME: 'WalletTabHome',
  },
};

export default Routes;
