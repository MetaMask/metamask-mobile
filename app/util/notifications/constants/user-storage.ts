export const USER_STORAGE_VERSION = '1';

export const USER_STORAGE_VERSION_KEY: unique symbol = 'v' as any;
