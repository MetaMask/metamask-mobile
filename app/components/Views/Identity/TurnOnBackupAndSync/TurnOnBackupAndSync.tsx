import React, { Fragment } from 'react';
import { Image, View, Linking, ScrollView } from 'react-native';

import Button, {
  ButtonVariants,
} from '../../../../component-library/components/Buttons/Button';
import { strings } from '../../../../../locales/i18n';
import Text, {
  TextColor,
  TextVariant,
} from '../../../../component-library/components/Texts/Text';
import { useTheme } from '../../../../util/theme';
import EnableBackupAndSyncCardImage from '../../../../images/enableBackupAndSyncCard.png';
import { createStyles } from './TurnOnBackupAndSync.styles';
import AppConstants from '../../../../core/AppConstants';
import SwitchLoadingModal from '../../../UI/Notification/SwitchLoadingModal';
import { useNavigation } from '@react-navigation/native';
import { useSelector } from 'react-redux';
import { RootState } from '../../../../reducers';
import { selectIsBackupAndSyncEnabled } from '../../../../selectors/identity';
import Routes from '../../../../constants/navigation/Routes';
import { useBackupAndSync } from '../../../../util/identity/hooks/useBackupAndSync';
import { BACKUPANDSYNC_FEATURES } from '@metamask/profile-sync-controller/user-storage';
import { MetaMetricsEvents, useMetrics } from '../../../hooks/useMetrics';
import { selectIsMetamaskNotificationsEnabled } from '../../../../selectors/notifications';

export const turnOnBackupAndSyncTestIds = {
  view: 'turn-on-backup-and-sync-view',
  cancelButton: 'turn-on-backup-and-sync-cancel-button',
  enableButton: 'turn-on-backup-and-sync-enable-button',
};

const TurnOnBackupAndSync = () => {
  const theme = useTheme();
  const styles = createStyles(theme);
  const navigation = useNavigation();

  const { setIsBackupAndSyncFeatureEnabled } = useBackupAndSync();
  const { trackEvent, createEventBuilder } = useMetrics();

  const isBasicFunctionalityEnabled = useSelector((state: RootState) =>
    Boolean(state?.settings?.basicFunctionalityEnabled),
  );
  const isBackupAndSyncEnabled = useSelector(selectIsBackupAndSyncEnabled);
  const isMetamaskNotificationsEnabled = useSelector(
    selectIsMetamaskNotificationsEnabled,
  );

  const goToLearnMore = () => {
    Linking.openURL(AppConstants.URLS.PROFILE_SYNC);
  };

  const trackEnableBackupAndSyncEvent = (newValue: boolean) => {
    trackEvent(
      createEventBuilder(MetaMetricsEvents.SETTINGS_UPDATED)
        .addProperties({
          settings_group: 'security_privacy',
          settings_type: 'profile_syncing',
          old_value: !newValue,
          new_value: newValue,
          was_notifications_on: isMetamaskNotificationsEnabled,
        })
        .build(),
    );
  };

  const handleGoBack = () => {
    navigation.goBack();
  };

  const handleEnableBackupAndSync = async () => {
    if (!isBasicFunctionalityEnabled) {
      navigation.navigate(Routes.MODAL.ROOT_MODAL_FLOW, {
        screen: Routes.SHEET.CONFIRM_TURN_ON_BACKUP_AND_SYNC,
        params: {
          enableBackupAndSync: async () => {
            navigation.navigate(Routes.SETTINGS_VIEW, {
              screen: Routes.SETTINGS.BACKUP_AND_SYNC,
            });
            await setIsBackupAndSyncFeatureEnabled(
              BACKUPANDSYNC_FEATURES.main,
              true,
            );
          },
          trackEnableBackupAndSyncEvent,
        },
      });
      return;
    }
    if (!isBackupAndSyncEnabled) {
      await setIsBackupAndSyncFeatureEnabled(BACKUPANDSYNC_FEATURES.main, true);
    }
    navigation.navigate(Routes.SETTINGS_VIEW, {
      screen: Routes.SETTINGS.BACKUP_AND_SYNC,
    });
  };

  return (
    <Fragment>
      <View style={styles.wrapper} testID={turnOnBackupAndSyncTestIds.view}>
        <Text
          variant={TextVariant.HeadingMD}
          color={TextColor.Default}
          style={styles.textTitle}
        >
          {strings('backupAndSync.enable.title')}
        </Text>
        <ScrollView>
          <View style={styles.card}>
            <Image source={EnableBackupAndSyncCardImage} style={styles.image} />
          </View>
          <Text
            variant={TextVariant.BodyMD}
            color={TextColor.Alternative}
            style={styles.textSpace}
          >
            {strings('backupAndSync.enable.description')}{' '}
            <Text
              variant={TextVariant.BodyMD}
              color={TextColor.Info}
              onPress={goToLearnMore}
            >
              {strings('backupAndSync.privacyLink')}
            </Text>
          </Text>
          <Text
            style={styles.textSettings}
            variant={TextVariant.BodyMD}
            color={TextColor.Alternative}
          >
            {strings('backupAndSync.enable.updatePreferences')}
          </Text>
          <Text variant={TextVariant.BodyMDBold}>
            {strings('backupAndSync.enable.settingsPath')}
          </Text>
        </ScrollView>

        <View style={styles.btnContainer}>
          <Button
            variant={ButtonVariants.Secondary}
            label={strings('notifications.activation_card.cancel')}
            onPress={handleGoBack}
            style={styles.ctaBtn}
            testID={turnOnBackupAndSyncTestIds.cancelButton}
          />
          <Button
            variant={ButtonVariants.Primary}
            label={strings('notifications.activation_card.cta')}
            onPress={handleEnableBackupAndSync}
            style={styles.ctaBtn}
            testID={turnOnBackupAndSyncTestIds.enableButton}
          />
        </View>
      </View>
      <SwitchLoadingModal
        loading={false}
        loadingText={strings('app_settings.enabling_notifications')}
      />
    </Fragment>
  );
};

export default TurnOnBackupAndSync;
