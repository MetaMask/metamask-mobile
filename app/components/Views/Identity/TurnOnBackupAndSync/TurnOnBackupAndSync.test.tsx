import React from 'react';

import TurnOnBackupAndSync, {
  turnOnBackupAndSyncTestIds,
} from './TurnOnBackupAndSync';
import renderWithProvider from '../../../../util/test/renderWithProvider';
import Routes from '../../../../constants/navigation/Routes';
import { fireEvent, waitFor } from '@testing-library/react-native';
import { BACKUPANDSYNC_FEATURES } from '@metamask/profile-sync-controller/user-storage';

const MOCK_STORE_STATE = {
  engine: {
    backgroundState: {
      UserStorageController: {
        isProfileSyncingEnabled: true,
      },
      AuthenticationController: {
        isSignedIn: true,
      },
    },
  },
  settings: {
    basicFunctionalityEnabled: true,
  },
};

const { InteractionManager } = jest.requireActual('react-native');

InteractionManager.runAfterInteractions = jest.fn(async (callback) =>
  callback(),
);

const mockNavigate = jest.fn();
jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockNavigate,
    }),
  };
});

const mockSetIsBackupAndSyncFeatureEnabled = jest.fn();
jest.mock('../../../../util/identity/hooks/useBackupAndSync', () => ({
  useBackupAndSync: () => ({
    setIsBackupAndSyncFeatureEnabled: mockSetIsBackupAndSyncFeatureEnabled,
    error: null,
  }),
}));

describe('TurnOnBackupAndSync', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const { toJSON } = renderWithProvider(<TurnOnBackupAndSync />, {
      state: MOCK_STORE_STATE,
    });
    expect(toJSON()).toMatchSnapshot();
  });

  it('enables backup and sync when clicking on the cta if backup and sync is disabled, and navigates to backup and sync settings either way', async () => {
    const { getByTestId } = renderWithProvider(<TurnOnBackupAndSync />, {
      state: {
        ...MOCK_STORE_STATE,
        engine: {
          backgroundState: {
            ...MOCK_STORE_STATE.engine.backgroundState,
            UserStorageController: {
              isProfileSyncingEnabled: false,
            },
          },
        },
      },
    });

    const switchElement = getByTestId(turnOnBackupAndSyncTestIds.enableButton);
    fireEvent.press(switchElement);

    await waitFor(() => {
      expect(mockSetIsBackupAndSyncFeatureEnabled).toHaveBeenCalledWith(
        BACKUPANDSYNC_FEATURES.main,
        true,
      );
      expect(mockNavigate).toHaveBeenCalledWith(Routes.SETTINGS_VIEW, {
        screen: Routes.SETTINGS.BACKUP_AND_SYNC,
      });
    });
  });

  it('opens a modal when clicking on the cta while basic functionality is off', () => {
    const { getByTestId } = renderWithProvider(<TurnOnBackupAndSync />, {
      state: {
        ...MOCK_STORE_STATE,
        engine: {
          backgroundState: {
            ...MOCK_STORE_STATE.engine.backgroundState,
            UserStorageController: {
              isProfileSyncingEnabled: false,
            },
          },
        },
        settings: {
          ...MOCK_STORE_STATE.settings,
          basicFunctionalityEnabled: false,
        },
      },
    });

    const switchElement = getByTestId(turnOnBackupAndSyncTestIds.enableButton);
    fireEvent.press(switchElement);

    expect(mockNavigate).toHaveBeenCalledWith(Routes.MODAL.ROOT_MODAL_FLOW, {
      screen: Routes.SHEET.CONFIRM_TURN_ON_BACKUP_AND_SYNC,
      params: {
        enableBackupAndSync: expect.any(Function),
        trackEnableBackupAndSyncEvent: expect.any(Function),
      },
    });
  });
});
