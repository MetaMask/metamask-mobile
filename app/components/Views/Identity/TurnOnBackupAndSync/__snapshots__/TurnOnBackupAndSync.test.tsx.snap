// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TurnOnBackupAndSync renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "flex-start",
      "backgroundColor": "#ffffff",
      "flex": 1,
      "paddingHorizontal": 16,
      "paddingTop": 60,
    }
  }
  testID="turn-on-backup-and-sync-view"
>
  <Text
    accessibilityRole="text"
    style={
      {
        "alignSelf": "center",
        "color": "#121314",
        "fontFamily": "CentraNo1-Bold",
        "fontSize": 18,
        "fontWeight": "700",
        "letterSpacing": 0,
        "lineHeight": 24,
        "marginBottom": 16,
      }
    }
  >
    Turn on backup and sync
  </Text>
  <RCTScrollView>
    <View>
      <View
        style={
          {
            "alignSelf": "center",
            "borderRadius": 12,
            "height": 240,
            "marginBottom": 32,
            "overflow": "hidden",
            "width": 718,
          }
        }
      >
        <Image
          source={1}
          style={
            {
              "height": 240,
              "resizeMode": "cover",
              "width": 718,
            }
          }
        />
      </View>
      <Text
        accessibilityRole="text"
        style={
          {
            "color": "#686e7d",
            "fontFamily": "CentraNo1-Book",
            "fontSize": 16,
            "fontWeight": "400",
            "letterSpacing": 0,
            "lineHeight": 24,
            "marginBottom": 16,
          }
        }
      >
        Backup and sync lets us store encrypted data for your custom settings and features. This keeps your MetaMask experience the same across devices and restores settings and features if you ever need to reinstall MetaMask.
         
        <Text
          accessibilityRole="text"
          onPress={[Function]}
          style={
            {
              "color": "#4459ff",
              "fontFamily": "CentraNo1-Book",
              "fontSize": 16,
              "fontWeight": "400",
              "letterSpacing": 0,
              "lineHeight": 24,
            }
          }
        >
          Learn how we protect your privacy
        </Text>
      </Text>
      <Text
        accessibilityRole="text"
        style={
          {
            "color": "#686e7d",
            "flexDirection": "column",
            "fontFamily": "CentraNo1-Book",
            "fontSize": 16,
            "fontWeight": "400",
            "letterSpacing": 0,
            "lineHeight": 24,
          }
        }
      >
        You can update your preferences at any time in
      </Text>
      <Text
        accessibilityRole="text"
        style={
          {
            "color": "#121314",
            "fontFamily": "CentraNo1-Bold",
            "fontSize": 16,
            "fontWeight": "700",
            "letterSpacing": 0,
            "lineHeight": 24,
          }
        }
      >
        Settings &gt; Backup and sync.
      </Text>
    </View>
  </RCTScrollView>
  <View
    style={
      {
        "alignSelf": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 16,
      }
    }
  >
    <TouchableOpacity
      accessibilityRole="button"
      accessible={true}
      activeOpacity={1}
      onPress={[Function]}
      onPressIn={[Function]}
      onPressOut={[Function]}
      style={
        {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "transparent",
          "borderColor": "#4459ff",
          "borderRadius": 12,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 40,
          "justifyContent": "center",
          "margin": 4,
          "paddingHorizontal": 16,
          "width": "48%",
        }
      }
      testID="turn-on-backup-and-sync-cancel-button"
    >
      <Text
        accessibilityRole="text"
        style={
          {
            "color": "#4459ff",
            "fontFamily": "CentraNo1-Medium",
            "fontSize": 16,
            "fontWeight": "500",
            "letterSpacing": 0,
            "lineHeight": 24,
          }
        }
      >
        Cancel
      </Text>
    </TouchableOpacity>
    <TouchableOpacity
      accessibilityRole="button"
      accessible={true}
      activeOpacity={1}
      onPress={[Function]}
      onPressIn={[Function]}
      onPressOut={[Function]}
      style={
        {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#4459ff",
          "borderRadius": 12,
          "flexDirection": "row",
          "height": 40,
          "justifyContent": "center",
          "margin": 4,
          "paddingHorizontal": 16,
          "width": "48%",
        }
      }
      testID="turn-on-backup-and-sync-enable-button"
    >
      <Text
        accessibilityRole="text"
        style={
          {
            "color": "#ffffff",
            "fontFamily": "CentraNo1-Medium",
            "fontSize": 16,
            "fontWeight": "500",
            "letterSpacing": 0,
            "lineHeight": 24,
          }
        }
      >
        Turn on
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;
